lib/streamlit/state/session_state.py

```diff
@@ -412,26 +412,52 @@ def reset_triggers(self) -> None:
     def cull_nonexistent(self, widget_ids: Set[str]):
         self._new_widget_state.cull_nonexistent(widget_ids)
 
+        # Remove entries from _old_state corresponding to widgets not in
+        # widget_ids that did *not* have their values initially set via
+        # st.session_state.
+        self._old_state = {
+            k: v
+            for k, v in self._old_state.items()
+            if (k in widget_ids or k not in self._initial_widget_values)
+        }
+
     def set_metadata(self, widget_metadata: WidgetMetadata) -> None:
         widget_id = widget_metadata.id
         self._new_widget_state.widget_metadata[widget_id] = widget_metadata
 
     def maybe_set_state_value(self, widget_id: str) -> bool:
+        """Keep widget_state and session_state in sync when a widget is registered.
+
+        This method returns whether the frontend needs to be updated with the
+        new value of this widget.
+        """
         widget_metadata = self._new_widget_state.widget_metadata[widget_id]
         deserializer = widget_metadata.deserializer
         initial_value = deserializer(None, widget_metadata.id)
 
         if widget_id not in self:
+            # This is the first time this widget is being registered, so we set
+            # its value in session_state and remember its initial_value.
```

Comment:
Is `_old_state[widget_id] = initial_value` what is referred to by "set it's value in initial state"?

Reply:
The `_old_state[widget_id] = initial_value` line sets the widget value to be accessible via `st.session_state` if a user does so shortly after (before we've received any updated widget state from the frontend), and the ` _initial_widget_values[widget_id] = initial_value` line remembers the initial value remembers what the widget's initial value is so that we can send the new value to the frontend in case it changed (which may happen if the user makes a code change to their script)